Rollup Node Spec (#43)
ethereum-optimism/optimistic-specs: rollup-node.md

[`ExecutionPayloadV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#executionpayloadv1

## Handling L1 Re-Orgs

The [previous section on L2 chain derivation][l2-chain-derivation] assumes
linear progression of the L1 chain. It is also applicable for batch processing,
meaning that any given point in time, the canonical L2 chain is given by
processing the whole L1 chain since the [L2 chain inception].

The L1 chain may occasionally re-organize, meaning the head of the L1 chain
changes to a block that is not the child of the previous head but rather one of
its "cousins" (i.e. the descendant of an ancestor of the previous head). In
those case, the rollup driver must:

1. Locate the *common ancestor*, a block that is an ancestor of both the
   previous and new head.
2. Isolate the range of L1 blocks \]common ancestor, ..., new head\].
3. For each such block, call [`engine_forkchoiceUpdatedOPV1`] and
   [`engine_getPayloadV1`].
   - Fill the [`PayloadAttributesOPV1`] object according to [the section on payload attributes][payload-attr].

>    - Fill the [``ForkchoiceStateV1`] object according to the [the section on the

Reviewer: ```suggestion
   - Fill the [`ForkchoiceStateV1`] object according to the [the section on the
```

Fix text formatting

     execution engine][calling-exec-engine], but set `headBlockHash` to the hash
     of the last processed L2 block (use the hash of the common ancestor
     initially) instead of the last L2 chain head. `safeBlockHash` and
     `finalizedBlockHash` must be updated accordingly.

> Note that post-[merge], the L1 chain will offer finalization guarantees
> meaning that it won't be able to re-org more than `FINALIZATION_DELAY_BLOCKS`
> in the past, hence preserving our finalization guarantees.

> **TODO** Enunciate our finalization guarantees.

> **TODO** Must enunciate error cases more precisely, as well as state what
> should be impossible and should cause the node to stop deriving the L2 chain.

## Block Gossip

Another responsability of the rollup node is to transmit the L2 blocks it

> derives from the L1 chain to other L2 nodes via a process called *block gossip*.

Author: missing glossary link for *block gossip*

[L2 derivation inputs]: /glossary.md#L2-derivation-inputs
[payload attributes]: /glossary.md#payload-attributes
[block]: /glossary.md#block
[execution engine]: /glossary.md#execution-engine
[reorg]: /glossary.md#re-organization
[block gossip]: /glossary.md#block-gossip
[rollup driver]: /glossary.md#rollup-driver
[deposits]: /glossary.md#deposits
[deposit-feed]: /glossary.md#L2-deposit-feed-contract
[L2 chain inception]: /glossary.md#L2-chain-inception

The [rollup node] is the component responsible for [deriving the L2
chain][derivation] from L1 blocks. This process happens in two steps:

1. Read [L2 derivation inputs] from L1 and generate [payload attributes] (essentially
   [a block without output properties][block]).
2. Pass the payload attributes to the [execution engine], so that [output block
   properties][block] may be computed.

While this process is conceptually a pure function from the L1 chain to the L2
chain, it is in practice incremental. The L2 chain changes whenever new L1
blocks (and hence new L2 derivation inputs) are added to the L1 chain, or
whenever the L1 chain [re-organizes][reorg].

Additionally, the rollup node is responsible for [block gossip], i.e.

> transmitting the L2 blocks it derives over a peer-to-peer network.

Reviewer: Same as other comment about p2p/transmitting: it is the job of the sequencer to seed the gossip, and unrelated to local derivation at the start, until the node figures out it either derives the same thing it already propagated, or derived something that will reorg away the previous sequencer claim.

The part of the rollup node that derives the L2 chain is called the [rollup
driver]. This document is mostly concerned with the specification of the rollup
driver, while block gossip is [specified in a separate document][gossip-spec].

[gossip-spec]: TODO

> **TODO LINK** block gossip spec

## L2 Chain Derivation
[l2-chain-derivation]: #l2-chain-derivation

> This section specifies how the [rollup driver] derives one L2 block from every

Reviewer: Are we fully committed to doing 1:1 derivation? I thought we would map 1 L1 block to 1 *or more* L2 blocks.

Author: It's 1-1 derivation because this is the deposits-only stage. It will change later on when we add sequencer blocks.

L1 block. The L2 block will carry the L1 block attributes (as a *[L1 attributes
transaction]*) well as all L2 transactions deposited by users in the L1 block
(*[deposits]*).

[L1 attributes transaction]: /glossary.md#l1-attributes-transaction

### From L2 derivation inputs to payload attributes

[payload-attr]: #From-L2-derivation-inputs-to-payload-attributes
[`PayloadAttributesOPV1`]: #From-L2-derivation-inputs-to-payload-attributes

The rollup reads the following [L2 derivation inputs] for each L1 block:

- L1 block attributes
   - block number
   - timestamp

>    - basefee

Reviewer: Not important now, but something we might want to add here later: creating that "logs root" (something more neat than the receipt root on L1). Merkleizing the L1 log events more efficiently, filtered to only those of the successful receipts, would be a nice thing to do as rollup. If L2 doesn't pay for a log-subscription in advance (the extension we talked about), at least it can efficiently provide a proof later on that something happened on L1.

Author: Could you write up your thoughts on that as an issue with a specific tag? Like `ideas`, `improvements` or something like that?

- [deposits]

A deposit is an L2 transaction that has been submitted on L1, via a transaction
sent to the [L2 deposit feed contract][deposit-feed].

While deposits are notably (but not only) used to "deposit" (bridge) ETH and
tokens to L2, the word *deposit* should be understood as "a transaction
*deposited* to L2".

The L1 attributes are read from the L1 block header, while deposits are read
from the block's log entries. Refer to the [**L2 deposit feed contract
specification**][deposit-feed] for details on how deposits are encoded as
log entries.

[deposit-feed-spec]: TODO

> **TODO LINK** deposit feed contract specification

From these L2 derivation inputs, the rollup node constructs an expanded version
of the [Engine API PayloadAttributesV1 object][PayloadAttributesV1]:

[PayloadAttributesV1]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payloadattributesv1

```

> PayloadAttributesOPV1: {

Reviewer: We should move this into its own block-derivation doc to keep this doc light enough to cover all rollup-node responsibilities.

Author: I don't get it, block derivation is 90% of the content of the present document, so you would move the whole document??

be sent over JSON-RPC. `array` refers to a JSON array.

[HEX value encoding]: https://eth.wiki/json-rpc/API#hex-value-encoding
[JSON-RPC-API]: https://eth.wiki/json-rpc/API

The object properties must be set as follows:

- `timestamp` is set to the current [unix time] (number of elapsed seconds since
  00:00:00 UTC on 1 January 1970), rounded to the closest multiple of 2 seconds.
- `random` is currently set to all zero bytes. After [the merge][merge], the randomness
  value exposed for the L1 block will be copied in this property.
- `suggestedFeeRecipient` is set to an address where the sequencer would like to
  direct the fees
- `transactions` is an array of transactions, RLP-encoded in the [EIP-2718]
  format (i.e. as a sequence of a single byte and a byte array), of the allowed
  types. We specify how to select the transactions below.

[unix type]: https://en.wikipedia.org/wiki/Unix_time
[merge]: https://ethereum.org/en/eth2/merge/
[EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718

> For details on RLP encoding, refer to the yellow paper, or look at [this
> hyperlinked implementation][encode-tx].

> [encode-tx]: https://github.com/norswap/nanoeth/blob/cc5d94a349c90627024f3cd629a2d830008fec72/src/com/norswap/nanoeth/transactions/Transaction.java#L84-L130

Reviewer: EIP-2718 defines all we need, we should not leak the details of other transaction types into this specification, or we will have to keep updating it with every L1 change. Also, like the other review comment, we should not rely on external implementations in this specification.

Author: Mmh, that is an interesting perspective indeed, I think I agree.

Regarding the impl link, the point is again to be informative, but your point is well-taken. Maybe we should try to mark it more explicitly? I really believe in the value-add here, the encoding scheme is dead simple, but it's hard to piece together from the yellowpaper (and actually the EIPs were most of them are specified & which I don't link here).

> [encode-tx]: https://github.com/norswap/nanoeth/blob/cc5d94a349c90627024f3cd629a2d830008fec72/src/com/norswap/nanoeth/transactions/Transaction.java#L84-L130

Reviewer 2: Perhaps link to geth's RLP package or ethers as well since its widely used?

Author: My goal with the link was to specify / explain how to reimplement rather than point to things that can be used. I must say RLP encoding is the thing that I looked up in Geth that confused me the most by far. This is probably being due to me not being super familiar with the language + this part of the code relying heavily on multiple kind of indirections and reflection, but I expect other people to trip on the same things. So for understanding purposes, it doesn't seem ideal.

I'm not strictly against linking libraries that could be used for implementation, but is this something we want to do in general?

> In particular, we allow the following two transaction types:

Reviewer: You are forgetting about the transaction type introduced in the Berlin hardfork (a subset of the EIP-1559 spec, supporting access-lists without the base-fee functionality)

Author: Didn't forget, was an explicit choice, but one that doesn't make a lot of sense, cf. https://github.com/ethereum-optimism/optimistic-specs/pull/43#discussion_r759679488

> In particular, we allow the following two transaction types:

Reviewer 2: Is it intentional to block type 1 transactions? The ones with the accesslist that aren't eip1559

Author: I figured it makes for easier spec and easier implementation, and shouldn't be an impediment at the wallet level (since they needed to update for us anyway, after EIP-1559).

That being said, maybe there are some legacy things we do absolutely wish to support or that will byte us in the ass if we don't?

I'm thinking that over the long run Ethereum is going to deprecate the non EIP-1559 transactions anyway (similarly to what is already underway with pre-EIP-155 transactions that are being soft-deprecated by Geth (there might be an EIP?)).

Author: On further reflection, I think it's a mistake to try to constrain the type of transactions that are allowed here. See proto's comments to that effect: https://github.com/ethereum-optimism/optimistic-specs/pull/43#discussion_r759678629

- [EIP-1559]: the current "regular" Ethereum L1 transaction type, which
  separates gas fees between a base fee and a tip
- the *[L1 attributes transaction]*, see below

[EIP-1559]: https://eips.ethereum.org/EIPS/eip-1559

Historical Optimistic Ethereum transactions also include [EIP-155] transactions,
which are not accepted by the network anymore. This is not an [EIP-2718] typed
transaction, but can be distinguished from those by the fact that their

> RLP-encoding's first byte will always be `>= 0xC0`.

Reviewer: Why? IMHO we should keep supporting it. Or we will have to modify geth to break support, which seems very backwards.

Author: Yeah, good point. I was thinking this was somehow simpler but I agree we should just follow L1 Ethereum.

[EIP-155]: https://eips.ethereum.org/EIPS/eip-155

#### Payload Transaction Format

The `transactions` array is filled with the deposits, prefixed by the (single)

> [L1 attributes transaction]. The deposits (which must be similarly RLP-encoded)

Reviewer: The typed-transaction EIP does not require us to RLP-encode the deposits if we don't want to. And it requires us to prefix it with a type byte. Let's keep the rollup-node doc general, and discuss the deposit format in a separate PR (@maurelian can lead this)

are simply copied byte-for-byte — it is the role of the [execution engine] to
reject invalidly-formatted transactions.

> **TODO** must offer some precisions on the format of deposits: sender,
> receivers both in-tx-as-encoded, and on-L2-tx. What about the fees?

The Optimistic Ethereum specific *[L1 attributes transaction]* has the following

> [EIP-2718]-compatible format: `0x7E || [block_number, timestamp, basefee]`

Reviewer: Wasn't @maurelian going to specify the deposits? We should separate this out, and keep it in the deposits specification doc.

Author: Yeah, this could be replaced by a link to that spec. Seemed like a simple enough strawman to put in.

Author: Actually revisiting this, this is wrong, this transaction does not appear at all on the L1 side since it is derived from L1 block attributes. Should probably be specified in the execution engine instead. Needs to be mentionned here too in any case, so duplication is warranted. We can change it later, obviously.

where:

- `0x7E` is the transaction type identifier. It is selected because transaction
  type identifiers are currently allowed to go up to `0x7F`. Picking a high
  identifier minimizes the risk that the identifier will be used by Ethereum in
  the future. We don't pick `0x7F` itself in case it becomes used for a
  variable-length encoding scheme.
- `block_number` is the L1 block number as a 64-bit integer (4 bytes)
- `timestamp` is the L1 block timestamp as a 64-bit integer (4 bytes)
- `basefee` is the L1 block basefee as a 64-bit integer (4 bytes)

When included in the `transactions` array, this transaction should be
RLP-encoded in the same way as other transactions.

> Here is an example valid `PayloadAttributesOPV1` object, which contains an L1

Reviewer: Let's move this into a doc specific to the execution-engine?

Author: What about having the person that writes the spec move it?

  ]
}
```

### Building the L2 block with the execution engine
[calling-exec-engine]: #building-the-L2-block-with-the-execution-engine

The Optimistic Ethereum [execution engine] is [specified in its own
document][execution-engine-spec].

This section defines how the rollup driver must interact with the execution
engine's in order to convert [payload attributes] into L2 blocks.

> **TODO** This section probably includes too much redundant details that will
> need to be removed once the execution engine spec is up.

Optimistic Ethereum's execution engine API is built upon [Ethereum's Engine API
specification][eth-engine-api], with a couple of modifications. That
specification builds upon [Ethereum's JSON-RPC API specification][JSON-RPC-API],
which itself builds upon the [JSON-RPC specification][JSON-RPC].

[eth-engine-api]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md
[JSON-RPC]: https://www.jsonrpc.org/specification

In particular, the [Ethereum's Engine API specification][eth-engine-api]

> specifies a [JSON-RPC] endpoint with a number of JSON-RPC (REST) routes, which

Reviewer: ```suggestion
specifies a [JSON-RPC] endpoint with a number of JSON-RPC routes, which
```
JSON-RPC is not REST. We kept the engine API design simple enough that it could be, but it's still JSON-RPC.

Author: Big philosophical debate, goal was just to give a quick idea, but not a hill that I'm willing to die on, will remove.

are the means through which the rollup driver interacts with the execution
engine.

> Instead of calling [`engine_forkchoiceUpdatedV1`], the rollup driver must call

Reviewer: I'm not sure if the two should be different. They can be made to be compatible:
- if `transactions` is empty, do the normal thing
- if `transactions` is not empty, do the optimism version (i.e. all the same, except pulling transactions into block from this list, instead of the tx pool)

And the `transactions` field itself is optional, defaults to an empty list.

Author: I'm just basing myself of the current version of the spec that doesn't have the `transactions` fields.
But if it changes, then I agree we should do it as you say.

the new [`engine_forkchoiceUpdatedOPV1`] route. This has the same signature,
except that:

[`engine_forkchoiceUpdatedV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#engine_forkchoiceupdatedv1
[`engine_forkchoiceUpdatedOPV1`]: TODO

> **TODO LINK** forkchoiceUpdatedOPV1 from spec

- it takes a [`PayloadAttributesOPV1`] object as input instead of
  [`PayloadAttributesV1`]. The execution engine must include the valid
  transactions supplied in this object in the block, in the same order as they
  were supplied, and only those. See the [previous section][payload-attr] for
  the specification of how the properties must be set.

- we repurpose the [`ForkchoiceStateV1`] structure with the following property
  semantics:
  - `headBlockHash`: block hash of the last block of the L2 chain, according to
    the rollup driver.

>   - `safeBlockHash`: same as `headBlockHash`.

Reviewer: This is changing on L1, and just a placeholder semantic, the two are not the same. And in the rollup they are definitely different. Also, this should be covered in the execution engine doc, not here in the general rollup-node doc.

Author: Yup, they're not the same but for the deposits-only stage we have nothing better to put in there, so I specified away.

I do think it belongs here too, because the rollup-node needs to fill in the arguments to make the call.

Reviewer 3: I agree that what's currently written here is fine. It's a placeholder for when there is a sequencer and thus `unsafe` is more meaningful.

The `payloadID` returned by [`engine_forkchoiceUpdatedOPV1`] can then be passed
to [`engine_getPayloadV1`] in order to obtain an [`ExecutionPayloadV1`], which
fully defines a new L2 block.

[`engine_getPayloadV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#engine_getpayloadv1
[`ExecutionPayloadV1`]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#executionpayloadv1

## Handling L1 Re-Orgs

The [previous section on L2 chain derivation][l2-chain-derivation] assumes
linear progression of the L1 chain. It is also applicable for batch processing,
meaning that any given point in time, the canonical L2 chain is given by
processing the whole L1 chain since the [L2 chain inception].

The L1 chain may occasionally re-organize, meaning the head of the L1 chain
changes to a block that is not the child of the previous head but rather one of
its "cousins" (i.e. the descendant of an ancestor of the previous head). In
those case, the rollup driver must:

1. Locate the *common ancestor*, a block that is an ancestor of both the
   previous and new head.

> 2. Isolate the range of L1 blocks \]common ancestor, ..., new head\].

Author: You're correct. I'm adding `code` markers. What would you replace it with? I'm not aware of a notation that would be as universal as this one.

> 3. For each such block, call [`engine_forkchoiceUpdatedOPV1`] and

Reviewer: This is closer to what manual sync would look like, but in some cases it may be better to first try a forkchoice update, and have the execution engine resolve the sync (like it would do in the L1 case).

Author: Mmh that is interesting. We should definitely talk about how things work on L1 here.

Author: No discussion on this, really agnostic on what it should be. There's still the chance that the upstream spec updates to accomodate our needs.

Author: Resolving this, this will be a concern for the execution engine.

   [`engine_getPayloadV1`].
   - Fill the [`PayloadAttributesOPV1`] object according to [the section on payload attributes][payload-attr].
   - Fill the [`ForkchoiceStateV1`] object according to the [the section on the
     execution engine][calling-exec-engine], but set `headBlockHash` to the hash
     of the last processed L2 block (use the hash of the common ancestor
     initially) instead of the last L2 chain head. `safeBlockHash` and
     `finalizedBlockHash` must be updated accordingly.

> Note that post-[merge], the L1 chain will offer finalization guarantees
> meaning that it won't be able to re-org more than `FINALIZATION_DELAY_BLOCKS`
> in the past, hence preserving our finalization guarantees.

> **TODO** Enunciate our finalization guarantees.

> **TODO** Must enunciate error cases more precisely, as well as state what
> should be impossible and should cause the node to stop deriving the L2 chain.

## Block Gossip

Another responsability of the rollup node is to transmit the L2 blocks it
derives from the L1 chain to other L2 nodes via a process called *[block
gossip]*.

> This is specified in [a separate document][gossip-spec].

Reviewer: When linking docs we should just use the document title, instead of "separate document" etc.

chain, it is in practice incremental. The L2 chain is extended whenever new L1
blocks (and hence new L2 derivation inputs) are added to the L1 chain.
Similarly, the L2 chain re-organizes whenever the L1 chain
[re-organizes][reorg].

The part of the rollup node that derives the L2 chain is called the [rollup
driver]. This document is currently only concerned with the specification of the
rollup driver.

## Table of Contents

- [L2 Chain Derivation][l2-chain-derivation]
  - [From L2 derivation inputs to payload attributes][payload-attr]
  - [Payload Transaction Format][payload-format]
  - [Building the L2 block with the execution engine][calling-exec-engine]
- [Handling L1 Re-Orgs][l1-reorgs]
- [Finalization Guarantees][finalization]

## L2 Chain Derivation
[l2-chain-derivation]: #l2-chain-derivation

This section specifies how the [rollup driver] derives one L2 block from every
L1 block. The L2 block will carry the L1 block attributes (as a *[L1 attributes
transaction]*) well as all L2 transactions deposited by users in the L1 block
(*[deposits]*).

[L1 attributes transaction]: /glossary.md#l1-attributes-transaction

> ### From L2 derivation inputs to payload attributes

Reviewer 4: ```suggestion
### From L1 blocks to L2 payloads
```

Is what I would suggest given top-level comment of this review.

[payload-attr]: #From-L2-derivation-inputs-to-payload-attributes

[`PayloadAttributesOPV1`]: #From-L2-derivation-inputs-to-payload-attributes

> The rollup reads the following [L2 derivation inputs] for each L1 block:

Reviewer 4: ```suggestion
The L2 chain derivation function will parse the following info from each L1 block:
```

Author: "L2 derivation inputs" are no more!

*deposited* to L2".

The L1 attributes are read from the L1 block header, while deposits are read
from the block's log entries. Refer to the [**deposit feed contract
specification**][deposit-feed] for details on how deposits are encoded as log
entries.

[deposit-feed-spec]: TODO

> **TODO LINK** deposit feed contract specification

From these L2 derivation inputs, the rollup node constructs an expanded version
of the [Engine API PayloadAttributesV1 object][PayloadAttributesV1]:

[PayloadAttributesV1]: https://github.com/ethereum/execution-apis/blob/main/src/engine/specification.md#payloadattributesv1

```
PayloadAttributesOPV1: {
	timestamp: QUANTITY
	random: DATA (32 bytes)
	suggestedFeeRecipient: DATA (20 bytes)
	transactions: array of DATA
}
```

The type notation used here refers to the [HEX value encoding] used by the
[Ethereum JSON-RPC API specification][JSON-RPC-API], as this structure will need to
be sent over JSON-RPC. `array` refers to a JSON array.

[HEX value encoding]: https://eth.wiki/json-rpc/API#hex-value-encoding
[JSON-RPC-API]: https://eth.wiki/json-rpc/API

The object properties must be set as follows:

> - `timestamp` is set to the current [unix time] (number of elapsed seconds since
>   00:00:00 UTC on 1 January 1970), rounded to the closest multiple of 2 seconds.

Reviewer 3: ```suggestion
- `timestamp` is set to the current [unix time], rounded to the closest multiple of 2 seconds.
```

I think this is a case where we can just link out to the reference without adding context.

Reviewer 3: I think its worth also indicating that the timestamps must be unique, ie. no two L2 blocks have the same one.

Author: Which reference would you link though? We're not copying this from L1 (1)! This is also different from the L1 spec (which doesn't round at 2s), and that spec is hard to link (because it's inside a pdf).

Also even if we reused the spec AND it was reusable, this is literally 10 words, redundancy seems worth it here.

(1) Although maybe we should at this stage - good question! In the context of future stages, do we want deposit timestamp == L1 timestamp or do we want to reflect our potential processing lag in the timestamp (imagining we go with the 1 block / 2s proposal of Karl)?

The [EIP-2718] transactions must have a transaction type that is valid on L1, or
be an *[L1 attributes transaction]* (see below).

#### Payload Transaction Format
[payload-format]: #payload-transaction-format

The `transactions` array is filled with the deposits, prefixed by the (single)
[L1 attributes transaction]. The deposits are simply copied byte-for-byte — it
is the role of the [execution engine] to reject invalidly-formatted
transactions.

> **TODO** must offer some precisions on the format of deposits: sender,
> receivers both in-tx-as-encoded, and on-L2-tx. What about the fees?

The Optimistic Ethereum specific *[L1 attributes transaction]* has the following
[EIP-2718]-compatible format: `0x7E || [block_number, timestamp, basefee]`
where:

- `0x7E` is the transaction type identifier. It is selected because transaction
  type identifiers are currently allowed to go up to `0x7F`. Picking a high
  identifier minimizes the risk that the identifier will be used by Ethereum in
  the future. We don't pick `0x7F` itself in case it becomes used for a

>   variable-length encoding scheme.

Reviewer 3: Is this referring to some discussion elsewhere about using 0x7F? 
Perhaps we should choose some address space closer to the middle of the range, which is a less likely schelling point than the beginning or end.

Author: No this is all me.

idk, we should do what seems likely. I think EIPs are likely to just pick in sequential orders.

But in the very unlikely case where they'd need a variable-length bytecode encoding, the traditional way to do that is to use the last byte in the range as the variable-length marker.

In any case, whatever we pick will be relevant to the design of future EIPs anyway, so I'm not too afraid, but I feel like this is both a bit safer & cleaner.

Author: Marking this resolved, feel free to open an issue if you think it warrants further discussion.
